process/datadog_checks/process/process.py

```diff
@@ -53,7 +55,53 @@
 }
 
 
+# taken verbatim from o'reilly python cookbook
+class ReadWriteLock:
+    """ A lock object that allows many simultaneous "read locks", but
+    only one "write lock." """
+
+    def __init__(self):
+        self._read_ready = threading.Condition(threading.Lock())
+        self._readers = 0
+
+    def acquire_read(self):
+        """ Acquire a read lock. Blocks only if a thread has
+        acquired the write lock. """
+        self._read_ready.acquire()
+        try:
+            self._readers += 1
+        finally:
+            self._read_ready.release()
+
+    def release_read(self):
+        """ Release a read lock. """
+        self._read_ready.acquire()
+        try:
+            self._readers -= 1
+            if not self._readers:
+                self._read_ready.notifyAll()
+        finally:
+            self._read_ready.release()
+
+    def acquire_write(self):
+        """ Acquire a write lock. Blocks until there are no
+        acquired read or write locks. """
+        self._read_ready.acquire()
+        while self._readers > 0:
+            self._read_ready.wait()
+
+    def release_write(self):
+        """ Release a write lock. """
+        self._read_ready.release()
+
+
 class ProcessCheck(AgentCheck):
+    # process list, shared among all instances
+    process_list = []
+    process_list_lock = ReadWriteLock()
+    last_proc_list_ts = 0
+    proc_list_cache_duration = DEFAULT_PROC_LIST_CACHE_DURATION
```

Comment:
I think you should delegate all those attributes to another object. This way you can use `self` normally everywhere. Something like:

```
class ProcessListCache(object):
    process_list = []
  ....

class ProcessCheck:
    process_cache = ProcessListCache()
```